csharp/src/Ice/IceSSL/SSLEngine.cs

```diff
@@ -260,12 +197,68 @@ internal void verifyPeer(ConnectionInfo info, string description)
         }
     }
 
+    internal SslClientAuthenticationOptions createClientAuthenticationOptions(
+        RemoteCertificateValidationCallback remoteCertificateValidationCallback,
+        string host)
+    {
+        return new SslClientAuthenticationOptions
+        {
+            ClientCertificates = _certs,
+            LocalCertificateSelectionCallback = (sender, targetHost, certs, remoteCertificate, acceptableIssuers) =>
+            {
+                if (certs == null || certs.Count == 0)
+                {
+                    return null;
+                }
+                else if (certs.Count == 1)
+                {
+                    return certs[0];
+                }
+
+                // Use the first certificate that match the acceptable issuers.
+                if (acceptableIssuers != null && acceptableIssuers.Length > 0)
+                {
+                    foreach (X509Certificate certificate in certs)
+                    {
+                        if (Array.IndexOf(acceptableIssuers, certificate.Issuer) != -1)
+                        {
+                            return certificate;
+                        }
+                    }
+                }
+                return certs[0];
+            },
+            RemoteCertificateValidationCallback = remoteCertificateValidationCallback,
+            TargetHost = host,
+            CertificateRevocationCheckMode = X509RevocationMode.NoCheck
+        };
+    }
+
+    internal SslServerAuthenticationOptions createServerAuthenticationOptions(
+        RemoteCertificateValidationCallback remoteCertificateValidationCallback)
+    {
+        // Get the certificate collection and select the first one.
+        X509Certificate2 cert = null;
+        if (_certs.Count > 0)
+        {
+            cert = _certs[0];
+        }
+
+        return new SslServerAuthenticationOptions
+        {
+            ServerCertificate = cert,
```

Comment:
Do we every construct an intermediate chain?

Reply:
No, we can look into adding support for this later. I understand that this requires loading from a PEM file, when loading PFX/P12 .NET only loads the first certificate.

Reply:
So as of 3.7 we only ever send the the server cert (unless .NET builds the chain automatically)?

Reply:
The ServerCertificateContext was not available with .NET Framework